062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+   It also doesn't support many additional features such as JBOD storage.
+   Because the initial releases were very simple, they had also very simple configuration.
+   And it was possible to configure them in the `.spec.kafka.config` section of the `Kafka` CR.
+   The next - not yet released - version of the Strimzi Quotas plugin brings many improvements.
+   More details about how the new version works can be found in the [Cluster Wide Volume Usage Quota Management](./047-cluster-wide-volume-usage-quota-management.md) proposal.
+   It operates at the level of a whole Kafka cluster and it has improved ability to prevent replicated messages to fill in the broker storage.
+   It also supports JBOD storage.
+   But it has also much more complicated configuration.
+   The plugin instances running in the different Kafka brokers communicate with each other using a Kafka topic.
+   And need therefore to connect to the Kafka brokers as clients as well.
+   Having the user configure this directly in the `.spec.kafka.config` section would be much more complicated as the user would need to configure the keystores and truststores, their passswords, bootstrap address etc.
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests to add a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow configuring different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The build-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - the Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
+* The other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quotas plugin - will be configured as `type: strimzi`.
```

Comment:
I don't think `strimzi` is the right name. If we wrote a new kind of quota plugin in the future what would we call it?

Reply:
I would agree with you Tom but at this point, if you are mentioning a "new kind of quota plugin", even the current quota plugin name `kafka-quotas-plugin` is not that good, or? Because we could have in Strimzi a new kind of Kafka-related quota plugin.